compiler/rustc_lint/src/noop_method_call.rs

```diff
@@ -0,0 +1,88 @@
+    ///
+    /// ```rust
+    /// # #![allow(unused)]
+    /// struct Foo;
+    /// let foo = &Foo;
+    /// let clone: &Foo = foo.clone();
+    /// ```
+    ///
+    /// {{produces}}
+    ///
+    /// ### Explanation
+    ///
+    /// Some method calls are noops meaning that they do nothing. Usually such methods
+    /// are the result of blanket implementations that happen to create some method invocations
+    /// that end up not doing anything. For instance, `Clone` is implemented on all `&T`, but
+    /// calling `clone` on a `&T` where `T` does not implement clone, actually doesn't do anything
+    /// as references are copy. This lint detects these calls and warns the user about them.
+    pub NOOP_METHOD_CALL,
+    Warn,
+    "detects the use of well-known noop methods"
+}
+
+declare_lint_pass!(NoopMethodCall => [NOOP_METHOD_CALL]);
+
+impl<'tcx> LateLintPass<'tcx> for NoopMethodCall {
+    fn check_expr(&mut self, cx: &LateContext<'tcx>, expr: &'tcx Expr<'_>) {
+        // We only care about method calls
+        if let ExprKind::MethodCall(..) = expr.kind {
+            // Get the `DefId` only when dealing with an `AssocFn`
+            if let Some((DefKind::AssocFn, did)) =
+                cx.typeck_results().type_dependent_def(expr.hir_id)
+            {
+                // Check that we're dealing with a trait method
+                if let Some(trait_id) = cx.tcx.trait_of_item(did) {
+                    // Check we're dealing with one of the traits we care about
+                    if ![sym::Clone, sym::Deref, sym::Borrow]
+                        .iter()
+                        .any(|s| cx.tcx.is_diagnostic_item(*s, trait_id))
+                    {
+                        return;
+                    }
+
+                    let substs = cx.typeck_results().node_substs(expr.hir_id);
+                    // We can't resolve on types that recursively require monomorphization,
+                    // so check that we don't need to perfom substitution
+                    if !substs.needs_subst() {
+                        let param_env = cx.tcx.param_env(trait_id);
+                        // Resolve the trait method instance
+                        if let Ok(Some(i)) = ty::Instance::resolve(cx.tcx, param_env, did, substs) {
+                            // Check that it implements the noop diagnostic
+                            if [
+                                sym::noop_method_borrow,
+                                sym::noop_method_clone,
+                                sym::noop_method_deref,
+                            ]
+                            .iter()
+                            .any(|s| cx.tcx.is_diagnostic_item(*s, i.def_id()))
+                            {
+                                let expr_span = expr.span;
+
+                                cx.struct_span_lint(NOOP_METHOD_CALL, expr_span, |lint| {
+                                    let message = "call to method that does nothing";
```

Comment:
I think it would be a good idea to mention the receiver type that the method is invoked on - e.g. ".clone() is called on a reference &SomeType, which does nothing".

Reply:
I'm not sure the receiver type is that important. In particular I'm especially worried about complicated types (e.g., complicated nested iterators). How about the following:

 "call to `.clone()` from the `Clone` trait which in this case does nothing and can be removed".

Reply:
I think the message may be very confusing to users, especially in cases like `&NotClone.clone()`. Could we at least display the 'outermost' part of the type (e.g `&_`)?

Reply:
I’m not sure how we would do that. We could hard code that everything we care about right now is a reference (but this may change in the future). Or what do you have in mind?

Reply:
Hardcoding it would be fine for now. Since this lint is about calling methods with a weird receiver, I think we need to say *something* about the receiver in the message.